web/text_layer_builder.js

```diff
@@ -209,15 +208,70 @@ class TextLayerBuilder {
       end.classList.add("active");
     });
 
-    div.addEventListener("mouseup", () => {
-      const end = div.querySelector(".endOfContent");
-      if (!end) {
-        return;
-      }
+    const reset = () => {
       if (typeof PDFJSDev === "undefined" || !PDFJSDev.test("MOZCENTRAL")) {
         end.style.top = "";
       }
       end.classList.remove("active");
+      selectingThroughMouse = false;
+      prevRange = null;
+    };
+    div.addEventListener("pointerup", reset);
+
+    // On touchscreen devices, when expaning/reducing a selection through the
+    // 'bubbles' that mark the selection end/start the browser will not fire
+    // any pointer event, because the user is interacting with browser-specific
+    // UI and not with the webpage itself.
+    // To workaround this, we listen to the 'selectionchange' event and check if
+    // the selection includes elements from this page. This has two drawbacks:
+    //  - We can only detect changes once they start happening. If the first
+    //    movement that the user does is to move one of the selection bubbles
+    //    to a blank part of the page, the selection will flicker before
+    //    that we can apply the fix.
+    //  - We cannot detect when the user stops interacting with the selection,
+    //    because there is no difference between "the user is still holding
+    //    their finger on the selection 'bubble'" and "the user is lifter their
+    //    finger and then pressed the 'bubble' again". Thus, the .endOfContent
+    //    element will remain active until when the user interacts with the
+    //    page again, triggering a pointerup event.
+    document.addEventListener("selectionchange", () => {
```

Comment:
Unfortunately you can only listen to `selectionchange` on `document` and not on a specific element. As currently written, this event listener will never be cleaned up and will leak even after that this text layer is removed from the document.

I was thinking of using `MutationObserver` to unregister the event when `div.textLayer` is removed, but maybe you already have some cleanup logic for when removing the `.textLayer`?

Reply:
We have a `cancel` method, which seems like the appropriate spot to remove this listener.